src/sema/expression.rs

```diff
@@ -1314,6 +1310,143 @@ fn coerce_number(
     })
 }
 
+fn number_literal(
+    loc: &pt::Loc,
+    integer: &str,
+    exp: &str,
+    ns: &Namespace,
+    factor: &BigInt,
+    diagnostics: &mut Vec<Diagnostic>,
+    resolve_to: ResolveTo,
+) -> Result<Expression, ()> {
+    let integer = BigInt::from_str(integer).unwrap();
+
+    let n = if exp.is_empty() {
+        integer
+    } else {
+        let base10 = BigInt::from_str("10").unwrap();
+
+        if let Some(exp) = exp.strip_prefix('-') {
+            if let Ok(exp) = u8::from_str(exp) {
+                let exp = base10.pow(exp);
+                let res = BigRational::new(integer, exp);
+
+                if res.is_integer() {
+                    res.to_integer()
+                } else {
+                    return Ok(Expression::RationalNumberLiteral(*loc, Type::Rational, res));
+                }
+            } else {
+                diagnostics.push(Diagnostic::error(
+                    *loc,
+                    format!("exponent '{}' too large", exp),
+                ));
+                return Err(());
+            }
+        } else if let Ok(exp) = u8::from_str(exp) {
+            let exp = base10.pow(exp);
+
+            integer.mul(exp)
+        } else {
+            diagnostics.push(Diagnostic::error(
+                *loc,
+                format!("exponent '{}' too large", exp),
+            ));
+            return Err(());
+        }
+    };
+
+    bigint_to_expression(loc, &n.mul(factor), ns, diagnostics, resolve_to)
+}
+
+fn rational_number_literal(
+    loc: &pt::Loc,
+    integer: &str,
+    fraction: &str,
+    exp: &str,
+    factor: &BigInt,
```

Comment:
What is the `factor`?

Reply:
I've renamed it unit and added a comment